### PR TITLE
Fixed #24917 -- Made admindocs show model methods with arguments

### PR DESCRIPTION
**Pull request by zanderle**

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: docs/ref/contrib/admin/admindocs.txt [hunk 1]
```diff
@@ -12,14 +12,18 @@ docstrings of models, views, template tags, and template filters for any app in
...
-primarily intended for documenting templates, template tags, and filters.
-For example, model methods that require arguments are purposefully omitted
-from the documentation because they can't be invoked from templates. The app
-can still be useful since it doesn't require you to write any extra
-documentation (besides docstrings) and is conveniently available from the
-:mod:`Django admin <django.contrib.admin>`.
+This way, Django admin documentation generator provides offline documentation
+for all custom and built-in template tags and template filters.
+
+.. admonition:: Note
```
Comment (timgraham) [inline docs/ref/contrib/admin/admindocs.txt]: I don't see why this was moved to a note box (by the way, there is `.. note::`), but anyway, you should have a ``.. versionchanged:: 1.9` note as well as a mention in the release notes. See our [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist).

File: tests/admin_docs/tests.py [hunk 2]
```diff
@@ -368,6 +386,31 @@ def test_model_docstring_renders_correctly(self):
...
+    def test_model_detail_default_show_more(self):
+        self.assertContains(
+            self.response,
+            '<a href="?show_more=0">Show less</a>',
+            html=True
+        )
+
+    def test_model_detail_show_less(self):
+        with captured_stderr() as self.docutils_stderr:
+            show_less_response = self.client.get(
```
Comment (timgraham) [inline tests/admin_docs/tests.py]: I think this would be a bit more readable:
```
url = reverse('...
show_less_response = self.client.get(url)
```

File: docs/ref/contrib/admin/admindocs.txt (hunk 1 above)
Comment (zanderle) [inline docs/ref/contrib/admin/admindocs.txt]: To be honest, can’t remember exactly xD Feels like I copied/paste it from somewhere.

Will fix it tomorrow, along with the other things missing.

File: tests/admin_docs/tests.py [hunk 3]
```diff
@@ -368,6 +386,31 @@ def test_model_docstring_renders_correctly(self):
         self.assertContains(self.response, body, html=True)
         self.assertContains(self.response, model_body, html=True)
 
+    def test_model_detail_title(self):
+        self.assertContains(
```
Comment (timgraham) [inline tests/admin_docs/tests.py]: one line is probably fine for these assertions (longer lines up to 119 chars are preferred if it helps readability)
Is this a missing test? It's a bit confusing to add it as part of this change. You could make it a separate commit.

**Comment by timgraham** [comment]
Instead of a separate page, maybe it would be better to have one page, but put the methods that take arguments in a separate table (listing them in a column named "Field" seems a bit misnamed)? It would also be helpful to list the argument names for those methods.

Please uncheck "Patch needs improvement" on the ticket when you make ther updates. Thanks!

**Comment by zanderle** [comment]
@timgraham thank you for all the comments! Will let you know when it's all ready.

**Comment by zanderle** [comment]
@timgraham I've updated this branch. I hope I didn't miss anything.

File: django/contrib/admindocs/templates/admin_doc/model_detail.html [hunk 4]
```diff
@@ -27,6 +27,7 @@ <h2 class="subhead">{{ summary }}</h2>
 
 {{ description }}
 
+<h3>Fields</h3>
```
Comment (timgraham) [inline django/contrib/admindocs/templates/admin_doc/model_detail.html]: add ", properties, and methods without arguments"?

File: docs/ref/contrib/admin/admindocs.txt [hunk 5]
```diff
@@ -12,14 +12,16 @@ docstrings of models, views, template tags, and template filters for any app in
 :setting:`INSTALLED_APPS` and makes that documentation available from the
 :mod:`Django admin <django.contrib.admin>`.
 
-You may, to some extent, utilize :mod:`~django.contrib.admindocs` to quickly
```
Comment (timgraham) [inline docs/ref/contrib/admin/admindocs.txt]: Could you revert this documentation change? I think it merely restores an older version of the text (at least I don't see any substantive changes).

File: docs/releases/1.9.txt [hunk 6]
```diff
@@ -151,6 +151,11 @@ Minor features
 * The time picker widget includes a '6 p.m' option for consistency of having
   predefined options every 6 hours.
 
+:mod:`django.contrib.admin`
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+* Admin Docs now also display model's methods that take arguments.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: I'd mimic what you wrote before, "The models section now includes methods that take arguments."

File: docs/releases/1.9.txt [hunk 7]
```diff
@@ -151,6 +151,11 @@ Minor features
 * The time picker widget includes a '6 p.m' option for consistency of having
   predefined options every 6 hours.
 
+:mod:`django.contrib.admin`
```
Comment (timgraham) [inline docs/releases/1.9.txt]: admindocs

**Comment by timgraham** [comment]
How about including keyword arguments as well? It's a bit confusing to see `get_next_by_pub_date` in the output but no arguments listed (it takes `**kwargs`). Please rebase your branch so it merges cleanly too.

File: django/contrib/admindocs/views.py [hunk 8]
```diff
@@ -254,11 +255,19 @@ def get_context_data(self, **kwargs):
                 verbose = func.__doc__
                 if verbose:
                     verbose = utils.parse_rst(utils.trim_docstring(verbose), 'model', _('model:') + opts.model_name)
-                fields.append({
```
Comment (timgraham) [inline django/contrib/admindocs/views.py]: I think it would be nice to have no output if there's no description instead of "None" I think something like `verbose = verbose or ''` would accomplish this.

**Comment by zanderle** [comment]
I think this is it now. Should I squash everything to a single commit?

**Comment by timgraham** [comment]
For a signature like `(self, baz, rox, foo='laa', *args, **kwargs)`, I'd say we could omit 'self' since that's not really an argument that you use. The output is missing `*args` and it would be nice to show kwarg defaults such as 'laa' if possible.

File: django/contrib/admindocs/templates/admin_doc/model_detail.html [hunk 9]
```diff
@@ -27,6 +27,7 @@ <h2 class="subhead">{{ summary }}</h2>
 
 {{ description }}
 
+<h3>Fields, properties, and methods without arguments</h3>
```
Comment (timgraham) [inline django/contrib/admindocs/templates/admin_doc/model_detail.html]: Add translation for headers

File: tests/admin_docs/tests.py [hunk 10]
```diff
@@ -261,6 +261,26 @@ def test_method_excludes(self):
         self.assertNotContains(self.response, "<td>set_status</td>")
         self.assertNotContains(self.response, "<td>save_changes</td>")
 
+    def test_methods_with_arguments(self):
+        """
+        Test that methods that take arguments are also displayed
```
Comment (timgraham) [inline tests/admin_docs/tests.py]: You may omit the "Tests that" prefix since all tests test things and just describe the expected behavior.

File: django/contrib/admindocs/templates/admin_doc/model_detail.html [hunk 11]
```diff
@@ -48,6 +49,30 @@ <h2 class="subhead">{{ summary }}</h2>
...
+    <th>{% trans 'Arguments' %}</th>
+    <th>{% trans 'Description' %}</th>
+</tr>
+</thead>
+<tbody>
+{% for method in methods|dictsort:"name" %}
+<tr>
+    <td>{{ method.name }}</td>
+    <td>{{ method.arguments|join:', ' }}</td>
+    <td>{{ method.verbose }}{% if method.help_text %} - {{ method.help_text|safe }}{% endif %}</td>
```
Comment (timgraham) [inline django/contrib/admindocs/templates/admin_doc/model_detail.html]: I don't think annotating a method with a "help_text" attribute is common -- that's mainly for model fields I think.

Comment (zanderle) [inline django/contrib/admindocs/templates/admin_doc/model_detail.html]: True. Ah, the dangers of copy-pasting code...

**Comment by zanderle** [comment]
@timgraham I've updated and improved this branch based on your last comments. Let me know if it needs anything else.

File: django/utils/inspect.py [hunk 12]
```diff
@@ -64,10 +103,31 @@ def func_accepts_kwargs(func):
     )
 
 
+def func_accepts_var_args(func):
+    if six.PY2:
+        # Not all callables are inspectable with getargspec, so we'll
+        # try a couple different ways but in the end fall back on assuming
+        # it is -- we don't want to prevent registration of valid but weird
```
Comment (timgraham) [inline django/utils/inspect.py]: this comment about "prevent registration" is a bit weird since it's related to the registration of signals which the other function is used by, but not this one. I think it would be a good idea to add some unit tests for these new functions so we don't need to debug them through the admindocs tests.

File: django/utils/inspect.py [hunk 13]
```diff
@@ -43,6 +43,45 @@ def get_func_args(func):
     ]
 
 
+def get_func_full_args(func):
+    """
+    Returns a list of argument name, default value tuples. If the
```
Comment (timgraham) [inline django/utils/inspect.py]: Use PEP 0257 verb style for new docstrings: "Return ..."

File: tests/admin_docs/tests.py [hunk 14]
```diff
@@ -206,7 +206,7 @@ class DefaultRoleTest(AdminDocsTestCase):
 
     def test_parse_rst(self):
         """
-        Tests that ``django.contrib.admindocs.utils.parse_rst`` uses
+        ``django.contrib.admindocs.utils.parse_rst`` should use
```
Comment (timgraham) [inline tests/admin_docs/tests.py]: Please keep unrelated changes like this in a separate commit when squashing.

**Comment by timgraham** [comment]
buildbot, add to whitelist.

**Comment by zanderle** [comment]
@timgraham should I add anything else?

**Comment by timgraham** [comment]
merged in f3dc173240d2a91c831b08dd6820e1d83322f3da and 29d52583e77b247aceeb2f514177a8a6c21ed786, thanks!
